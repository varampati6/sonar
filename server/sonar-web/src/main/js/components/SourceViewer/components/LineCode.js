/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import classNames from 'classnames';
import LineIssuesList from './LineIssuesList';
import LocationIndex from '../../common/LocationIndex';
import LocationMessage from '../../common/LocationMessage';
import { splitByTokens, highlightSymbol, highlightIssueLocations } from '../helpers/highlight';
import type { Tokens } from '../helpers/highlight';
import type { SourceLine } from '../types';
import type { LinearIssueLocation } from '../helpers/indexing';
import type { Issue } from '../../issue/types';

type Props = {|
  highlightedLocationMessage?: { index: number, text: string },
  highlightedSymbols?: Array<string>,
  issues: Array<Issue>,
  issueLocations: Array<LinearIssueLocation>,
  line: SourceLine,
  onIssueChange: Issue => void,
  onIssueSelect: (issueKey: string) => void,
  onLocationSelect?: number => void,
  onSymbolClick: Array<string> => void,
  scroll?: HTMLElement => void,
  secondaryIssueLocations: Array<{
    from: number,
    to: number,
    line: number,
    index: number,
    startLine: number
  }>,
  selectedIssue: string | null,
  showIssues: boolean
|};

type State = {
  tokens: Tokens
};

export default class LineCode extends React.PureComponent {
  activeMarkerNode: ?HTMLElement;
  codeNode: HTMLElement;
  props: Props;
  state: State;
  symbols: NodeList<HTMLElement>;

  constructor(props: Props) {
    super(props);
    this.state = {
      tokens: splitByTokens(props.line.code || '')
    };
  }

  componentDidMount() {
    this.attachEvents();
    if (this.props.highlightedLocationMessage && this.activeMarkerNode && this.props.scroll) {
      this.props.scroll(this.activeMarkerNode);
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.line.code !== this.props.line.code) {
      this.setState({
        tokens: splitByTokens(nextProps.line.code || '')
      });
    }
  }

  componentWillUpdate() {
    this.detachEvents();
  }

  componentDidUpdate(prevProps: Props) {
    this.attachEvents();
    if (
      this.props.highlightedLocationMessage &&
      prevProps.highlightedLocationMessage !== this.props.highlightedLocationMessage &&
      this.activeMarkerNode &&
      this.props.scroll
    ) {
      this.props.scroll(this.activeMarkerNode);
    }
  }

  componentWillUnmount() {
    this.detachEvents();
  }

  attachEvents() {
    this.symbols = this.codeNode.querySelectorAll('.sym');
    for (const symbol of this.symbols) {
      symbol.addEventListener('click', this.handleSymbolClick);
    }
  }

  detachEvents() {
    if (this.symbols) {
      for (const symbol of this.symbols) {
        symbol.removeEventListener('click', this.handleSymbolClick);
      }
    }
  }

  handleSymbolClick = (e: Object) => {
    e.preventDefault();
    const keys = e.currentTarget.className.match(/sym-\d+/g);
    if (keys.length > 0) {
      this.props.onSymbolClick(keys);
    }
  };

  renderMarker(index: number, message: ?string) {
    const { onLocationSelect } = this.props;
    const onClick = onLocationSelect ? () => onLocationSelect(index) : undefined;
    const ref = message != null ? node => (this.activeMarkerNode = node) : undefined;
    return (
      <LocationIndex key={`marker-${index}`} onClick={onClick} selected={message != null}>
        <span href="#" ref={ref}>{index + 1}</span>
        {message != null && <LocationMessage selected={true}>{message}</LocationMessage>}
      </LocationIndex>
    );
  }

  render() {
    const {
      highlightedLocationMessage,
      highlightedSymbols,
      issues,
      issueLocations,
      line,
      onIssueSelect,
      secondaryIssueLocations,
      selectedIssue,
      showIssues
    } = this.props;

    let tokens = [...this.state.tokens];

    if (highlightedSymbols) {
      highlightedSymbols.forEach(symbol => {
        tokens = highlightSymbol(tokens, symbol);
      });
    }

    if (issueLocations.length > 0) {
      tokens = highlightIssueLocations(tokens, issueLocations);
    }

    if (secondaryIssueLocations) {
      tokens = highlightIssueLocations(tokens, secondaryIssueLocations, 'issue-location');

      if (highlightedLocationMessage) {
        const location = secondaryIssueLocations.find(
          location => location.index === highlightedLocationMessage.index
        );
        if (location) {
          tokens = highlightIssueLocations(tokens, [location], 'selected');
        }
      }
    }

    const className = classNames('source-line-code', 'code', {
      'has-issues': issues.length > 0
    });

    const renderedTokens = [];
    tokens.forEach((token, index) => {
      if (token.markers.length > 0) {
        token.markers.forEach(marker => {
          const message = highlightedLocationMessage != null &&
            highlightedLocationMessage.index === marker
            ? highlightedLocationMessage.text
            : null;
          renderedTokens.push(this.renderMarker(marker, message));
        });
      }
      renderedTokens.push(<span className={token.className} key={index}>{token.text}</span>);
    });

    return (
      <td className={className} data-line-number={line.line}>
        <div className="source-line-code-inner">
          <pre ref={node => (this.codeNode = node)}>{renderedTokens}</pre>
        </div>
        {showIssues &&
          issues.length > 0 &&
          <LineIssuesList
            issues={issues}
            onIssueChange={this.props.onIssueChange}
            onIssueClick={onIssueSelect}
            selectedIssue={selectedIssue}
          />}
      </td>
    );
  }
}
