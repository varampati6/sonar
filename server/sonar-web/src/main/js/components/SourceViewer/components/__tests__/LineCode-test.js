/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import LineCode from '../LineCode';

it('render code', () => {
  const line = {
    line: 3,
    code: '<span class="k">class</span> <span class="sym sym-1">Foo</span> {'
  };
  const issueLocations = [{ from: 0, to: 5, line: 3 }];
  const wrapper = shallow(
    <LineCode
      highlightedSymbols={['sym1']}
      issues={[{ key: 'issue-1' }, { key: 'issue-2' }]}
      issueLocations={issueLocations}
      line={line}
      onIssueSelect={jest.fn()}
      onSelectLocation={jest.fn()}
      onSymbolClick={jest.fn()}
      selectedIssue="issue-1"
      showIssues={true}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
