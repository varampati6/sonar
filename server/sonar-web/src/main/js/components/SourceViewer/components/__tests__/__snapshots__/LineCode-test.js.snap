// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render code 1`] = `
<td
  className="source-line-code code has-issues"
  data-line-number={3}
>
  <div
    className="source-line-code-inner"
  >
    <pre>
      <span
        className="k source-line-code-issue"
      >
        class
      </span>
      <span
        className=""
      >
         
      </span>
      <span
        className="sym sym-1"
      >
        Foo
      </span>
      <span
        className=""
      >
         {
      </span>
    </pre>
  </div>
  <LineIssuesList
    issues={
      Array [
        Object {
          "key": "issue-1",
        },
        Object {
          "key": "issue-2",
        },
      ]
    }
    onIssueClick={[Function]}
    selectedIssue="issue-1"
  />
</td>
`;
